### ISSUE TITLE
WARN The `redirect` prop on <Screen /> is deprecated and will be removed. Please use `router.redirect` 

### ISSUE THREAD

**Issue by gaetan224** [expo/router#834]
### Which package manager are you using? (Yarn is recommended)

yarn

### Summary

from the example from [expo router docs ](https://docs.expo.dev/router/advanced/stack/)
when running the for dynamic route

```import { View, Text } from 'react-native';
import { Stack, useLocalSearchParams, useRouter } from 'expo-router';

export default function Details() {
  const router = useRouter();
  const params = useLocalSearchParams();

  return (
    <View style={{ flex: 1, alignItems: 'center', justifyContent: 'center' }}>
      <Stack.Screen
        options={{
          title: params.name,
        }}
      />
      <Text
        onPress={() => {
          router.setParams({ name: 'Updated' });
        }}>
        Update the title
      </Text>
    </View>
  );
}```


i ahev the following error

 WARN  The `redirect` prop on <Screen /> is deprecated and will be removed. Please use `router.redirect` instead
    at Screen (http://192.168.1.6:8081/index.ts.bundle//&platform=ios&dev=true&hot=false&lazy=true:146799:20)
    at RCTView
    at View (http://192.168.1.6:8081/index.ts.bundle//&platform=ios&dev=true&hot=false&lazy=true:40431:43)
    at View



### Minimal reproducible example

import { View, Text } from 'react-native';
import { Stack, useLocalSearchParams, useRouter } from 'expo-router';

export default function Details() {
  const router = useRouter();
  const params = useLocalSearchParams();

  return (
    <View style={{ flex: 1, alignItems: 'center', justifyContent: 'center' }}>
      <Stack.Screen
        options={{
          title: params.name,
        }}
      />
      <Text
        onPress={() => {
          router.setParams({ name: 'Updated' });
        }}>
        Update the title
      </Text>
    </View>
  );
}

**Comment by marklawlor** [comment]
Please provide a full reproduction. Creating a new project `npx create-expo-app@latest --template tabs@49` and updating `app/(tabs)/index` to the "Minimal reproducible example" does not reproduce the issue.

I'm going to close this issue, but if you can provide information we can reopen it again.

**Comment by sync** [comment]
I had a similar issue on my side and what caused it was this code:

```
<Stack.Screen
        options={{
          title: params.name,
        }}
      />
```

If you remove it the warning will go away

**Comment by sync** [comment]
I think issue is here perhaps https://github.com/expo/router/blob/f69e31c84a83ceb6d678dfb8aeed708974d301a8/packages/expo-router/src/views/Screen.tsx#L49

if you look at the code, if you  do this for example:

 ```
<Stack.Screen
        options={{
          title: params.name,
        }}
      />
```

Then deprecated would end up being `undefined` because `redirect` would not be set

```
    useDeprecated(
      "The `redirect` prop on <Screen /> is deprecated and will be removed. Please use `router.redirect` instead",
      redirect
    );
```

Then looking at `useDepecated` code here: 

```
export function useDeprecated(
  message: string,
  guard: unknown = true,
  key = message
) {
  return useWarnOnce(key, guard, `Expo Router: ${message}`);
}
```
the `guard` would end up being set to `true` because of the default value `guard: unknown = true`

So for anyone using stack.screen we would end up with a warning, should this issue be reopened ?

**Comment by sync** [comment]
This code was introduced with the `2.0.1` release, by default @49 template is using expo-router 2.0.0

**Comment by sync** [comment]
will send a pr

**Comment by pdandradeb** [comment]
> Unless I am completely wrong, anyone using stack.screen would end up with a warning, should this issue be reopened ?

@sync yes, I can confirm that just using `<Stack.Screen options={{ title: '' }} />` without `redirect` triggers this warning.

Reverting to `2.0.0` "fixed" it.

**Comment by sync** [comment]
yes waiting for someone to review this: https://github.com/expo/expo/pull/23932 :-)

**Comment by zackify** [comment]
also confirmed this exists, thanks for the pr

**Comment by animaonline** [comment]
When will a fix be available?

**Comment by animaonline** [comment]
When can we expect a fix?

**Comment by Dundyne** [comment]
Same issue here, pretty annoying

**Comment by Pcummings** [comment]
Agree - let's get it fixed

**Comment by sahiljaidka224** [comment]
Can confirm this is still a problem, v 2.0.5

**Comment by matthewtesimale** [comment]
Same! Still annoying. It wasn't showing from a project I spun up a few days ago... But now it is showing.

**Comment by Pcummings** [comment]
> on expo router `2.0.5` I'm also seeing the issue. I was declaring screen options inline:
> 
> 
> 
> `return (
> 
>    <Stack.Screen options={{ headerShown: true, title: "ELLO PUPPET" }} />
> 
>    ...
> 
> )
> 
> `
> 
> 
> 
> Prior to upgrading, I was able to declare my screen options like this in my component, that is no longer the case
> 
> 
> 
> After removing this, and setting the screen options in the declared stack, the warning went away. 
> 
> 

Can you show an example of how you do that?

**Comment by roiiiu** [comment]
In my case, make sure all your router dictionary has an "index. tsx" file as a placeholder, and then the warning will gone.
<img width="214" alt="image" src="https://github.com/expo/router/assets/11494827/5abe83f0-1b36-47f6-a962-dea92c2e2042">
<img width="371" alt="image" src="https://github.com/expo/router/assets/11494827/46763b12-14b3-4f18-9058-8ffaa13c8041">

**Comment by roiiiu** [comment]
> > > on expo router `2.0.5` I'm also seeing the issue. I was declaring screen options inline:
> > > `return ( <Stack.Screen options={{ headerShown: true, title: "ELLO PUPPET" }} /> ... ) `
> > > Prior to upgrading, I was able to declare my screen options like this in my component, that is no longer the case
> > > After removing this, and setting the screen options in the declared stack, the warning went away.
> > 
> > 
> > Can you show an example of how you do that?
> 
> in `(auth)/_layout.tsx`
> 
> ```
> import { Stack } from "expo-router";
> import React from "react";
> 
> export default function AuthLayout() {
>   return (
>     <Stack>
>       <Stack.Screen
>         name="login"
>         options={{ headerShown: true, title: "Login" }}
>       />
>     </Stack>
>   );
> }
> ```

solved!

**Comment by riolly** [comment]
This will fix:
 ```diff
- <Stack.Screen options={{ title: "Home Page" }} />
+ <Stack screenOptions={{ title: "Home Page" }} />
```

**Comment by joy7426** [comment]
Remove this chunk of code from the screen

```typescript
     <Stack.Screen screenOptions={{ headerShown: false }} />
``` 

Now you can use this screen configuration in _layout.

**Comment by edmbn** [comment]
> Remove this chunk of code from the screen
> 
> ```ts
> <Stack.Screen screenOptions={{ headerShown: false }} />
> ```
> 
> Now you can use this screen configuration in _layout.

This shouldn't be the solution since sometimes there is logic you want to apply from the screen. For example, you may want to modify the `headerRight` component based on some conditions on the screen (loading data, etc.).

**Comment by ycldz** [comment]
This feature is very useful to use within the page and should not be removed.

**Comment by ycldz** [comment]
> > Remove this chunk of code from the screen
> > ```ts
> > <Stack.Screen screenOptions={{ headerShown: false }} />
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Now you can use this screen configuration in _layout.
> 
> This shouldn't be the solution since sometimes there is logic you want to apply from the screen. For example, you may want to modify the `headerRight` component based on some conditions on the screen (loading data, etc.).
> 
> https://docs.expo.dev/router/advanced/stack/#header-buttons

I agree, it is very useful to use it within the page, this feature should not be removed.

**Comment by joy7426** [comment]
> > Remove this chunk of code from the screen
> > ```ts
> > <Stack.Screen screenOptions={{ headerShown: false }} />
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Now you can use this screen configuration in _layout.
> 
> This shouldn't be the solution since sometimes there is logic you want to apply from the screen. For example, you may want to modify the `headerRight` component based on some conditions on the screen (loading data, etc.).
> 
> https://docs.expo.dev/router/advanced/stack/#header-buttons

I agree but nothing to do! They should have kept both in place.
However, I made my own header and perhaps this is the only solution.

**Comment by dhruvr4** [comment]
When I do  <Stack>
      <Stack.Screen
        name="login"
        options={{ headerShown: true, title: "Login" }}
      />
    </Stack> or  <Stack screenOptions = {{ headerShown: true, title: "Login" }} 
the warning goes away but the name of the path to the screen for example account/login comes written at the top of the screen

Does anyone know why or how to remove that

**Comment by amerllica** [comment]
Is this warning a critical issue or can we ignore it?

**Comment by danstepanov** [comment]
The solution, although incorrect in its implementation by the Expo team, is to move your declarations to `_layout.tsx`.  @marklawlor correct me if wrong here, I have a reproducible example and can show you the error occurring using npx create-expo-app as well create-expo-stack (more minified version).

```
return (
  <Stack>
    <Stack.Screen name="one" options={{ title: "One" }} />
    <Stack.Screen name="two"
        options={{ title: "Two", headerLeft: () => <BackButton /> }}
      />
  </Stack>
);
```

Moving forward, Expo Router should either continue to support the ability to specify Stack.Screen (as it did in v2.0.0, i believe) on the page itself (as this is often required and react-navigation supports it, so expo-router should) or Expo should communicate the new proper syntax for how to configure navigation options as you used to be able to via `<Stack.Screen />, on an individual screen.

**Comment by danstepanov** [comment]
> Is this warning a critical issue or can we ignore it?

you can ignore it until this is removed in future version of expo router, my assumption is that it won't be removed until v3 but I could be wrong.

**Comment by danstepanov** [comment]
The solution, although confusing, is to move your declarations to `_layout.tsx`.  @marklawlor correct me if wrong here, I have a reproducible example and can show you the error occurring using npx create-expo-app as well create-expo-stack (more minified version).

```
return (
  <Stack>
    <Stack.Screen name="one" options={{ title: "One" }} />
    <Stack.Screen name="two"
        options={{ title: "Two", headerLeft: () => <BackButton /> }}
      />
  </Stack>
);
```

It seems meaningful for Expo Router to continue to support the ability to specify Stack.Screen (as it did in v2.0.0, i believe) on a per screen basis (as this is often required and react-navigation supports it, so expo-router should). I'm curious why the Expo team opted to remove/change this?

**Comment by monarcode** [comment]
Having this issue right now and its driving me crazy, hope it gets fixed soon

**Comment by osmanyz** [comment]
The same thing is happing to me, it doesn't make sense at all. Are there any good news ?

**Comment by zackify** [comment]
Just use patch package to do the fix from above: https://github.com/expo/router/issues/834#issuecomment-1681340570

**Comment by monarcode** [comment]
> Just use patch package to do the fix from above: https://github.com/expo/router/issues/834#issuecomment-1681340570

How can we use this fix?

**Comment by nabilfreeman** [comment]
I'm on 2.0.8 and this error is still happening. Has the fix been deployed yet?

**Comment by edmbn** [comment]
> I'm on 2.0.8 and this error is still happening. Has the fix been deployed yet?
> 
> 
> 
> Update: This is NOT fixed - the PR that resolves this is currently open:
> 
> 
> 
> https://github.com/expo/router/pull/859

If you read the changelog it is marked in resolved in some version around 3.1. This WARNING (not error) is just happening in the current version.

**Comment by NiklasEi** [comment]
> If you read the changelog it is marked as resolved in some version around 3.1. This WARNING (not error) is just happening in the current version.

The fix is merged for version 3 in the expo/expo repository. The linked PR is the backport of that fix for version 2, which is the "latest" version on npmjs.

The fact that this is "just" a warning doesn't help here. It causes a pop-up in developer mode and spams my console.

**Comment by markedwards** [comment]
Why can this not be backported to a 2.0.x release?  This is incredibly annoying.

**Comment by mcsrk** [comment]
At this point lets just switch to react nav

**Comment by adekorir** [comment]
frustrating.

**Comment by Brian-McBride** [comment]
How else do you change the header components in a dynamic view?

`app/(shop)/(product)/[id].js` as an example.

I have 
```js
  <Stack.Screen
    options={{
      title: data?.name ?? '',
    }}
  />
```
Producing the same error as everyone else. 

I just need to set the title of my view based on the data that I've fetched.

**Comment by leanderlee** [comment]
Can this issue be reopened? This seems important to fix.

**Comment by zavbala** [comment]
Re-open it, please 🙃

**Comment by leanderlee** [comment]
> Can this issue be reopened? This seems important to fix.

pinging @marklawlor, in case.

**Comment by leggomuhgreggo** [comment]
Here's a different workaround that has the advantage of maintaining the "page + options" colocation.

```diff
import React from 'react';
import { View} from 'react-native';
import { Link, Stack } from 'expo-router';
- import { Stack } from 'expo-router';
+ import { useNavigation } from 'expo-router';

export const HomeScreen = () => {
+  const navigation = useNavigation();
+
+  React.useEffect(() => {
+    navigation.setOptions({
+      headerShown: false,
+    });
+  }, [navigation]);

  return (
    <View>
-      <Stack.Screen options={{ headerShown: false }} />

      <Text>Home</Text>
    </View>
  );
};
```

- [`useNavigation`](https://docs.expo.dev/router/reference/hooks/#usenavigation)
- [`setOptions`](https://reactnavigation.org/docs/navigation-prop#setoptions)


If that doesn't work whatever reason, the other workaround is is to move the `<Stack.Screen options={...}>` to the parent `<Stack />` in `_layout.tsx` (see this [comment](https://github.com/expo/router/issues/834#issuecomment-1720392215) and this [comment](https://github.com/expo/router/issues/834#issuecomment-1732850512))

Hope that helps!

**Comment by zackify** [comment]
If you want to keep doing stuff the old way in v2, you can patch it yourself. Someone asked me above how to do it so here's how:

just install [patch-package](https://www.npmjs.com/package/patch-package) from npm, or use the built in approach if your package manager supports it.

All you have to do is add `!= null` in the file referenced in the merged pr: https://github.com/expo/expo/pull/23932 to get this to go away

made a short [blog post](https://open.substack.com/pub/zachcodes/p/patching-npm-modules-with-patch-package?r=f5mmx&utm_campaign=post&utm_medium=web) on it if anyone needs more help than that.

**Comment by wvq** [comment]
> Here's a different workaround that has the advantage of maintaining the "page + options" colocation.
> 
> ```diff
> import React from 'react';
> import { View} from 'react-native';
> import { Link, Stack } from 'expo-router';
> - import { Stack } from 'expo-router';
> + import { useNavigation } from 'expo-router';
> 
> export const HomeScreen = () => {
> +  const navigation = useNavigation();
> +
> +  React.useEffect(() => {
> +    navigation.setOptions({
> +      headerShown: false,
> +    });
> +  }, [navigation]);
> 
>   return (
>     <View>
> -      <Stack.Screen options={{ headerShown: false }} />
> 
>       <Text>Home</Text>
>     </View>
>   );
> };
> ```
> 
> * [`useNavigation`](https://docs.expo.dev/router/reference/hooks/#usenavigation)
> * [`setOptions`](https://reactnavigation.org/docs/navigation-prop#setoptions)
> 
> If that doesn't work whatever reason, the other workaround is is to move the `<Stack.Screen options={...}>` to the parent `<Stack />` in `_layout.tsx` (see this [comment](https://github.com/expo/router/issues/834#issuecomment-1720392215) and this [comment](https://github.com/expo/router/issues/834#issuecomment-1732850512))
> 
> Hope that helps!
> 
> EDIT: Opened [PR](https://github.com/expo/expo/pull/24676) to update `<Stack>` [docs](https://docs.expo.dev/router/advanced/stack/#configure-header-bar) with example

With this example, there has a display issue on my cell phone. 
During the screen navigation animation, header is show, then hidden.
Use `<Screen options={{ headerShown: false }}>` the header is default hidden.

**Comment by lesthoward** [comment]
> The solution, although confusing, is to move your declarations to `_layout.tsx`. @marklawlor correct me if wrong here, I have a reproducible example and can show you the error occurring using npx create-expo-app as well create-expo-stack (more minified version).
> 
> ```
> return (
>   <Stack>
>     <Stack.Screen name="one" options={{ title: "One" }} />
>     <Stack.Screen name="two"
>         options={{ title: "Two", headerLeft: () => <BackButton /> }}
>       />
>   </Stack>
> );
> ```
> 
> It seems meaningful for Expo Router to continue to support the ability to specify Stack.Screen (as it did in v2.0.0, i believe) on a per screen basis (as this is often required and react-navigation supports it, so expo-router should). I'm curious why the Expo team opted to remove/change this?

You are right. I did the same thing to solve this issue, and you did the same thing as me. So, yes, this work!

**Comment by balloman** [comment]
Why is this issue closed??

**Comment by markedwards** [comment]
> I'm going to close this issue, but if you can provide information we can reopen it again.

@marklawlor, this is easily reproducible, can you please reopen?

**Comment by mattphillips** [comment]
As a workaround, while we wait for this to be patched in v2 I've found that you can suppress the warning by adding the following to the root of your app:

```ts
import { LogBox } from 'react-native';

LogBox.ignoreLogs([
  'The `redirect` prop on <Screen /> is deprecated and will be removed. Please use `router.redirect` instead'
]);
```

Not the best solution but saves faffing with patch package etc.

**Comment by tholst** [comment]
fyi: it seems to be fixed now in `expo-router  v2.0.9`

(see commit https://github.com/expo/router/commit/9f8a67be23de9e3d5e70400ab4dff590da4c1ff8)

**Comment by david7378** [comment]
> sup

Yes its fixed!! Thanks!!
